wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataProvider.kt

```diff
@@ -102,20 +116,185 @@ class WalletTransactionMetadataProvider @Inject constructor(
+                    )
+                } else if (metadata.rate == null && tx.exchangeRate != null) {
+                    val exchangeRate = tx.exchangeRate!!
+                    setTransactionExchangeRate(
+                        tx.txId,
+                        ExchangeRate(
+                            exchangeRate.fiat.currencyCode,
+                            exchangeRate.fiat.value.toString()
+                        )
+                    )
+                }
+
+                // sync transaction memo
+                if (metadata.memo.isNotBlank() && tx.memo == null) {
+                    tx.memo = metadata.memo
+                } else if (metadata.memo.isBlank() && tx.memo != null) {
+                    setTransactionMemo(tx.txId, tx.memo!!)
+                }
+            } else {
+                // it does not exist, so import everything from the transaction
+                log.info("sync transaction metadata not exists: ${tx.txId}")
+                insertTransactionMetadata(tx.txId)
+            }
+
+        }
+    }
+
     override suspend fun getTransactionMetadata(txId: Sha256Hash): TransactionMetadata? {
-        val transactionMetadata = transactionMetadataDao.load(txId)
+        var transactionMetadata = transactionMetadataDao.load(txId)
         if (transactionMetadata == null) {
             insertTransactionMetadata(txId)
         }
-        return transactionMetadataDao.load(txId)
+
+        transactionMetadata = transactionMetadataDao.load(txId)
+        if (transactionMetadata != null && transactionMetadata.taxCategory == null) {
+            transactionMetadata.taxCategory = getDefaultTaxCategory(txId)
+        }
+        return transactionMetadata
+    }
+
+    /**
+     * obtains the default tax category for transfers
+     */
+    private suspend fun getDefaultTaxCategory(txId: Sha256Hash): TaxCategory? {
+        var taxCategory: TaxCategory? = null
+        val tx = walletData.wallet!!.getTransaction(txId)
+        tx?.run {
+            if (tx.getValue(walletData.wallet!!).isNegative) {
+                // outgoing transaction, check inputs
+                for (input in inputs) {
+                    if (input.connectedOutput != null) {
+                        val output = input.connectedOutput!!
+                        val address = when {
+                            ScriptPattern.isP2PKH(output.scriptPubKey) ->
+                                Address.fromPubKeyHash(
+                                    walletData.networkParameters,
+                                    ScriptPattern.extractHashFromP2PKH(output.scriptPubKey)
+                                )
+                            ScriptPattern.isP2SH(output.scriptPubKey) ->
+                                Address.fromScriptHash(
+                                    walletData.networkParameters,
+                                    ScriptPattern.extractHashFromP2SH(output.scriptPubKey)
+                                )
```

Comment:
this probably would never be triggered because our wallets don't support P2SH addresses internally. 

Our wallet can send to P2SH addresses, which is why we can check for this below.